partiql-ast/src/main/java/org/partiql/ast/v1/AstVisitor.java

```diff
@@ -33,206 +32,557 @@
-
-    R visitExclude(Exclude node, C ctx);
-
-    R visitExcludePath(ExcludePath node, C ctx);
-
-    R visitExcludeStep(ExcludeStep node, C ctx);
-
-    R visitExcludeStepStructField(ExcludeStep.StructField node, C ctx);
-
-    R visitExcludeStepCollIndex(ExcludeStep.CollIndex node, C ctx);
-
-    R visitExcludeStepStructWildcard(ExcludeStep.StructWildcard node, C ctx);
-
-    R visitExcludeStepCollWildcard(ExcludeStep.CollWildcard node, C ctx);
-
-    R visitFrom(From node, C ctx);
-
-    R visitTableRef(FromTableRef node, C ctx);
-
-    R visitFromExpr(FromExpr node, C ctx);
-
-    R visitFromJoin(FromJoin node, C ctx);
-
-    R visitLet(Let node, C ctx);
-
-    R visitLetBinding(Let.Binding node, C ctx);
-
-    R visitGroupBy(GroupBy node, C ctx);
-
-    R visitGroupByKey(GroupBy.Key node, C ctx);
-
-    R visitOrderBy(OrderBy node, C ctx);
-
-    R visitSort(Sort node, C ctx);
-
-    R visitSetOp(SetOp node, C ctx);
-
-    R visitGraphMatch(GraphMatch node, C ctx);
-
-    R visitGraphMatchPattern(GraphPattern node, C ctx);
-
-    R visitGraphPart(GraphPart node, C ctx);
-
-    R visitGraphPartNode(GraphPart.Node node, C ctx);
-
-    R visitGraphPartEdge(GraphPart.Edge node, C ctx);
-
-    R visitGraphPartPattern(GraphPart.Pattern node, C ctx);
-
-    R visitGraphQuantifier(GraphQuantifier node, C ctx);
-
-    R visitGraphSelector(GraphSelector node, C ctx);
-
-    R visitGraphSelectorAnyShortest(GraphSelector.AnyShortest node, C ctx);
-
-    R visitGraphSelectorAllShortest(GraphSelector.AllShortest node, C ctx);
-
-    R visitGraphSelectorAny(GraphSelector.Any node, C ctx);
-
-    R visitGraphSelectorAnyK(GraphSelector.AnyK node, C ctx);
-
-    R visitGraphSelectorShortestK(GraphSelector.ShortestK node, C ctx);
-
-    R visitGraphSelectorShortestKGroup(
-            GraphSelector.ShortestKGroup node,
-            C ctx
-    );
-
-    R visitGraphLabel(GraphLabel node, C ctx);
-
-    R visitGraphLabelName(GraphLabel.Name node, C ctx);
-
-    R visitGraphLabelWildcard(GraphLabel.Wildcard node, C ctx);
-
-    R visitGraphLabelNegation(GraphLabel.Negation node, C ctx);
-
-    R visitGraphLabelConj(GraphLabel.Conj node, C ctx);
-
-    R visitGraphLabelDisj(GraphLabel.Disj node, C ctx);
+public abstract class AstVisitor<R, C> {
+    public R defaultVisit(AstNode node, C ctx) {
+        for (AstNode child : node.children()) {
+            child.accept(this, ctx);
+        }
+        return defaultReturn(node, ctx);
+    }
+
+    public abstract R defaultReturn(AstNode node, C ctx);
+
+    public R visit(AstNode node, C ctx) {
+        return node.accept(this, ctx);
+    }
+
+    public R visitStatement(Statement node, C ctx) {
+        if (node instanceof Query) {
+            return visitQuery((Query) node, ctx);
+        } else if (node instanceof Explain) {
+            return visitExplain((Explain) node, ctx);
+        } else {
+            throw new IllegalStateException("Unexpected value: " + node);
+        }
```

Comment:
I'm trying to think about how we can make it easier for users to add their own custom nodes without breaking their visitors completely. Is there something wrong with just:
```suggestion
        node.accept(this, ctx)
```

Reply:
Theoretically, each variant of a sum type would look like:
```java
public class Query : AstNode() {
    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitQuery(this, ctx);
    }
}
```

So, it would definitely work for the built-in variants. For the other variants, I'm not entirely sure how they'd be able to invoke a custom visit (AKA `visitor.visitCustomNode(this, ctx)`).

Reply:
You're right about the simplification. Those sum type visit functions can be simplified to `node.accept(this, ctx)`.

Reply:
From offline discussion, there's pros/cons to both approaches. There are some drawbacks to the `node.accept` default for the sum types, primarily a user of this library would need to be careful with overriding the proper methods when adding their own variants. This can hopefully be resolved by including some good usage guides.

I do like your suggested approach a bit more since it will simplify what we (the PartiQL team) need to add when adding a new variant.